src/compiler/checker.ts

```diff
@@ -6244,6 +6254,41 @@ namespace ts {
             }
         }
 
+        function typeIdenticalToSomeType(source: Type, target: UnionOrIntersectionType): boolean {
+            for (let t of target.types) {
+                if (isTypeIdenticalTo(source, t)) {
+                    return true;
+                }
+            } 
+            return false;
+        }
+
+        /**
+         * Return the reduced form of the source type. This type is computed by by removing all source
+         * constituents that have an identical match in the target type.
+         */
+        function reduceUnionOrIntersectionType(source: UnionOrIntersectionType, target: UnionOrIntersectionType) {
+            let sourceTypes = source.types;
+            let sourceIndex = 0;
+            let modified = false;
+            while (sourceIndex < sourceTypes.length) {
+                if (typeIdenticalToSomeType(sourceTypes[sourceIndex], target)) {
+                    if (!modified) {
+                        sourceTypes = sourceTypes.slice(0);
+                        modified = true;
+                    }
+                    sourceTypes.splice(sourceIndex, 1);

```

Comment:
I don't think I understand the mutability discipline for the checker. It looks like `splice` modifies `source.types`. Is this OK? How do we know that nobody else refers to `source`? (even though `inferFromTypes` no longer does after the call to `reduceUnionOrIntersection`.)